server/tserver/src/main/java/org/apache/accumulo/tserver/ThriftClientHandler.java

```diff
@@ -686,16 +685,27 @@ private void setUpdateTablet(UpdateSession us, KeyExtent keyExtent) {
 
   @Override
   public void applyUpdates(TInfo tinfo, long updateID, TKeyExtent tkeyExtent,
-      List<TMutation> tmutations) {
+      List<TMutation> tmutations) throws TException {
     UpdateSession us = (UpdateSession) server.sessionManager.reserveSession(updateID);
     if (us == null) {
       return;
     }
 
+    ThreadPoolExecutor writeThreadExecutor = server.getWriteThreadExecutor();
     boolean reserved = true;
+
     try {
       KeyExtent keyExtent = KeyExtent.fromThrift(tkeyExtent);
-      setUpdateTablet(us, keyExtent);
+
+      // get write thread pool handler but only for user tablets
+      if (TabletType.type(keyExtent) == TabletType.USER) {
+        if (writeThreadExecutor.isTerminating()) {
+          throw new TException("Write Thread pool is terminating");
+        }
+        writeThreadExecutor.execute(() -> setUpdateTablet(us, keyExtent));
```

Comment:
I'm not sure that there is any benefit to performing `setUpdateTablet` in another thread. The method either queues up the mutations for the Tablet, or it add them all to the failure queue. Also, I believe this has to be done before you get to the code on line 712 below. Otherwise there may be no queuedMutations.

Reply:
That is a good point, I don't think there is any benefit from using a separate thread. I was just using a `ThreadPoolExecutor` since that is what was returned from the ThreadPools method I found. I wasn't sure how to use `ThreadPools` to manager a thread pool.

Reply:
Is there a benefit to running `setUpdateTablet` in a different thread, then having to wait for it to finish to move on?

Reply:
I don't know. It might help if we knew the exact bottle neck. `setUpdateTablet()` is just populating the `UpdateSession`. There is another method after that flushes the update session.

This makes me wonder if the original problem was just too many mutations going to a single tablet.

Reply:
I'm not sure there's a benefit to doing setUpdateTablet in a separate thread. I don't think that's where the action is happening. I think the bulk of the action is happening further down in this method.

Before any of this code is changed, I think one could argue that each of these threads that is handling the Thrift request are "write threads". We can't really limit the number of threads that are created, because that's a function of the RPC handling. However, I think the idea was to limit the number of these threads that are doing meaningful work by:

1. determine if the thread is a user thread, and if so, then,
2. perform the entire rest of the work to handle this RPC request in a separate size-limited executor service, while this RPC thread waits for it to complete before returning to the client.

So, if I understand this code correctly, it wouldn't just be calling setUpdateTablet in a separate thread... but doing all the work that this method is doing in a separate thread.

Honestly, though, I'm not sure what the initial problem was that led to this feature request. I don't know what the bottleneck was for having all of these current RPC threads behaving as "write threads" without a limit.